components/sup/src/event/nats_message_stream.rs

```diff
@@ -4,93 +4,151 @@ use crate::event::{Error,
 use futures::{channel::{mpsc as futures_mpsc,
                         mpsc::UnboundedSender},
               stream::StreamExt};
-use rants::{error::Error as RantsError,
-            native_tls::TlsConnector,
-            Client,
-            Subject};
-use tokio::time;
+use nats::{self,
+           asynk::{Options, Connection}};
+use tokio::{
+    time, 
+    sync::Mutex};
+use std::{
+    io,
+    path::PathBuf,
+    sync::Arc};
 
 /// The subject and payload of a NATS message.
 #[derive(Debug)]
 pub struct NatsMessage {
-    subject: &'static Subject,
+    subject: &'static str,
     payload: Vec<u8>,
 }
 
 impl NatsMessage {
-    pub fn new(subject: &'static Subject, payload: Vec<u8>) -> Self {
+    pub fn new(subject: &'static str, payload: Vec<u8>) -> Self {
         NatsMessage { subject, payload }
     }
 
     pub fn payload(&self) -> &[u8] { self.payload.as_slice() }
 }
 
+#[derive(Clone)]
+struct NatsClientImpl {
```

Comment:
You can refactor the NatsClient to avoid needing the NatsClientImpl and the extra `Arc<Mutex>`. You can try:
```
#[derive(Clone)]
struct NatsClient {
    connection: Arc<Mutex<Option<Connection>>>,
}

impl NatsClient {
    fn new() -> NatsClient {
        NatsClient { connection: Arc::new(Mutex::new(None)) }
    }

    async fn connect(self, supervisor_id: String, config: EventStreamConfig) -> io::Result<()> {        
        match config.connect_method.into() {
            Some(timeout) => {
                trace!("Timeout used -> {:?}", timeout);
                
                time::timeout(timeout, self.connect_impl(supervisor_id, &config))
                    .await??;
            } 
            None => {
                trace!("Timeout not used");
                tokio::spawn(async move { self.connect_impl(supervisor_id, &config).await });
            }
        }
        Ok(())
    }

    async fn connect_impl(self, supervisor_id: String, config: &EventStreamConfig) -> io::Result<()> {        
        if self.connection.lock().await.is_none() {
            match Self::options_from_config(&supervisor_id, config) 
                .connect(&config.url.to_string())
                .await {
                    Ok(conn) => {
                        *self.connection.lock().await = Some(conn);
                    }
                    Err(e) => {
                        Error::ConnectNatsServer(e);
                    }
                }
        }
        Ok(())
    }

    fn options_from_config(supervisor_id: &str, config: &EventStreamConfig) -> Options {
        let name = format!("hab_client_{}", supervisor_id);
        let ca_certs = match habitat_core::tls::native_tls_wrapper::installed_cacerts(None) {
            Ok(ca_certs) => ca_certs,
            Err(_)         => None
        };
        match config.server_certificate {
            Some(ref nats_options) => {
                let cert_path: PathBuf = nats_options.clone().into(); 
                Options::with_token(&config.token.to_string())
                    .with_name(&name)
                    .add_root_certificate(cert_path)
                    .add_root_certificate(ca_certs.unwrap())
                    .with_retry_on_failed_connect()
            }
            None => {
                Options::with_token(&config.token.to_string()) 
                    .with_name(&name)
                    .add_root_certificate(ca_certs.unwrap())
                    .with_retry_on_failed_connect()
            }
        }
    }

    async fn publish(&self, subject: &str, msg: impl AsRef<[u8]>) -> io::Result<()> {
        if let Some(conn) = &*self.connection.lock().await {
            conn.publish(subject, msg).await
        }
        else {
            Err(io::Error::new(io::ErrorKind::Other, "Not connected to NATS server!"))
        }
    }
}
```

Reply:
I agree and like how this is evolving into less code.